spark/agent/src/main/scala/za/co/absa/spline/harvester/AttributeDependencySolver.scala

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright 2019 ABSA Group Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package za.co.absa.spline.harvester
+
+import java.util.UUID
+
+import za.co.absa.spline.harvester.ModelConstants.OperationParams._
+import za.co.absa.spline.harvester.ModelConstants.OperationExtras._
+import za.co.absa.spline.model.expr.{Alias, AttrRef, Expression, Generic}
+import za.co.absa.spline.producer.model.{DataOperation, OperationLike}
+
+import scala.collection.mutable
+
+
+case class AttributeDependencies(attributes: Map[UUID, Set[UUID]], operations: Map[UUID, Set[Int]])
+
+object AttributeDependencySolver {
+
+  def resolveDependencies(dataOperations: Seq[DataOperation], ioOperations: Seq[OperationLike]):
+      AttributeDependencies = {
+
+    val attributeDependencies = dataOperations
+      .map(resolveDependencies)
+      .reduce(mergeDependencies)
+
+    val operationDependencies = resolveAttributeOperationDependencies(
+      dataOperations ++ ioOperations,
+      attributeDependencies)
+
+    AttributeDependencies(attributeDependencies, operationDependencies)
+  }
+
+  private def resolveDependencies(op: DataOperation): Map[UUID, Set[UUID]] =
+    op.extra(Name) match {
+      case "Project" => resolveExpressionList(op.params(Transformations), op.schema)
+      case "Aggregate" => resolveExpressionList(op.params(Aggregations), op.schema)
+      case "Generate" => resolveGenerator(op)
+      case _ => Map.empty
+    }
+
+  private def resolveExpressionList(exprList: Any, maybeSchema: Option[Any]): Map[UUID, Set[UUID]] = {
+
+    val schema = maybeSchema.get.asInstanceOf[Seq[UUID]]
+
+    exprList
+      .asInstanceOf[Some[Seq[Expression]]]
+      .get
+      .zip(schema)
+      .map{ case (expr, attrId) => attrId -> toAttrDependencies(expr) }
+      .toMap
+  }
+
+  private def resolveGenerator(op: DataOperation): Map[UUID, Set[UUID]] = {
+
+    val expression = op.params.get("generator").asInstanceOf[Some[Generic]].get
+    val dependencies = toAttrDependencies(expression)
+
+    val keyId = op.params("generatorOutput")
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Usage of head on collections.](https://app.codacy.com/app/ABSA_OSS/spline/pullRequest?prid=4630741)